### PR TITLE
[Stacked, 1/2] Adds PhysicsNeMo-Mesh Tessellation utilities

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

Adds tools for tesselation (currently, triangulation; later, tetrahedralization and so on) to PhysicsNeMo-Mesh.

This is borne out of an ask from the HiLiftAeroML pre-processing in PhysicsNeMo-Curator.

Specifically, existing approaches to triangulation use fan-triangulation, which produces incorrect results for non-convex polygons. Non-convex polygons are a rare occurrence in CFD meshes, but they can appear in other cases.

This PR provides a clean, `torch-compile`-compatible (where possible) approach to tessellation. It also provides `Mesh.from_polygons()` which is a nice `@classmethod`-level constructor that reduces downstream duplicate code in PhysicsNeMo-Curator when doing direct imports from "polygon-soup" meshes (e.g., hand-importing from VTP).

This PR is Part 1 of a stacked PR; Part 2 will add a new Jupyter notebook with usage examples.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1711/commits/928358629eed3559236274bb5521be121e6e27c1

File: docs/api/mesh/core.rst [hunk 1]
```diff
@@ -19,6 +19,9 @@ available both as ``Mesh`` methods and as standalone functions in the
 corresponding submodules. The methods are thin wrappers that pass ``self`` to
 the standalone functions.
 
+To construct a triangle mesh from a non-simplicial polygon soup, use
+:meth:`Mesh.from_polygons` (see :doc:`tessellation`).
```
Comment (megnvidia) [inline docs/api/mesh/core.rst]:
```suggestion
:meth:`Mesh.from_polygons` (refer to :doc:`tessellation`).
```

File: docs/api/mesh/index.rst [hunk 2]
```diff
@@ -180,6 +180,9 @@ Key Features
   tangent spaces
 - **Mesh operations**: subdivision (linear, Loop, Butterfly), smoothing,
   remeshing, repair
+- **Tessellation**: triangulate polygon soups into simplicial meshes (convex
+  fan with an ear-clip fallback for non-convex polygons), e.g. via
```
Comment (megnvidia) [inline docs/api/mesh/index.rst]:
```suggestion
  fan with an ear-clip fallback for non-convex polygons), for example, using
```

File: docs/api/mesh/tessellation.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Decompose non-simplicial cells into the simplices that the
+:class:`~physicsnemo.mesh.mesh.Mesh` data structure stores. Currently this
+provides polygon-soup triangulation (:func:`triangulate_polygons`): a
+vectorized vertex-0 fan for convex polygons and `ear clipping
+<https://en.wikipedia.org/wiki/Polygon_triangulation>`_ for the rare non-convex
+ones.
+
+Handling non-convex polygons correctly matters for any unsigned-area-weighted
+quantity (wall-shear / viscous force integration, total wetted area): the
```
Comment (megnvidia) [inline docs/api/mesh/tessellation.rst]:
```suggestion
quantity (wall-shear / viscous force integration, or total wetted area): the
```

File: docs/api/mesh/tessellation.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+signed *vector* area of a vertex-0 fan telescopes to the polygon's regardless
```
Comment (megnvidia) [inline docs/api/mesh/tessellation.rst]:
```suggestion
signed *vector* area of a vertex-0 fan telescopes to the polygon's total area regardless
```

File: docs/api/mesh/tessellation.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+of convexity, but the sum of *unsigned* triangle areas does not.
```
Comment (megnvidia) [inline docs/api/mesh/tessellation.rst]:
```suggestion
of convexity, but the sum of the *unsigned* triangle areas does not.
```

File: docs/api/mesh/tessellation.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+Every ``k``-gon yields exactly ``k - 2`` triangles, so per-polygon data is
+broadcast to the output identically in both paths via the returned
```
Comment (megnvidia) [inline docs/api/mesh/tessellation.rst]:
```suggestion
broadcast to the output identically in both paths using the returned
```

File: physicsnemo/mesh/tessellation/_ear_clipping.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+# limitations under the License.
+
+"""Vectorized ear clipping for the rare non-convex polygons.
+
+A vertex-0 fan is only valid for convex polygons; for a non-convex polygon its
+fan triangles overlap and exit the polygon, so their unsigned areas no longer
+sum to the polygon's. Ear clipping produces a true triangulation (`k - 2`
+triangles, no added points) instead.
+
+Ear clipping is sequential per polygon, but we vectorize across polygons by
```
Comment (megnvidia) [inline physicsnemo/mesh/tessellation/_ear_clipping.py]:
```suggestion
Ear clipping is sequential per polygon, but is vectorized across polygons by
```

File: physicsnemo/mesh/tessellation/_ear_clipping.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+grouping them by valence ``k`` (the ``torch.unique`` group-by-count pattern used
+in :mod:`physicsnemo.mesh.calculus._neighborhoods`): every polygon in a group is
+clipped in lockstep over a static ``k - 2`` rounds. This is the rare fallback,
```
Comment (megnvidia) [inline physicsnemo/mesh/tessellation/_ear_clipping.py:27]:
```suggestion
clipped in lockstep over static ``k - 2`` rounds. This is the rare fallback,
```

File: physicsnemo/mesh/tessellation/triangulate.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+
+def _convex_mask(
+    points: Float[torch.Tensor, "n_points 3"],
+    polygons: Adjacency,
+    normals: Float[torch.Tensor, "n_polygons 3"],
+) -> Int[torch.Tensor, " n_polygons"]:
+    """Boolean mask, ``True`` where a polygon is convex (or degenerate).
+
+    A polygon is convex iff no vertex is reflex. We measure the signed sine of
```
Comment (megnvidia) [inline physicsnemo/mesh/tessellation/triangulate.py]:
```suggestion
    A polygon is convex iff no vertex is reflex. The signed sine of
```

File: physicsnemo/mesh/tessellation/triangulate.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    each vertex turn relative to the polygon normal (scale-free, in
```
Comment (megnvidia) [inline physicsnemo/mesh/tessellation/triangulate.py]:
```suggestion
    each vertex turn is measured relative to the polygon normal (scale-free, in
```

**Review by megnvidia — COMMENTED** [review]
I don't normally comment on .py files, but since the comment text was long, I did this time. Feel free to ignore that.  But, the use of 'we' is discouraged by our Nivida style guide.

File: physicsnemo/mesh/tessellation/_ear_clipping.py (hunk 8 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/tessellation/_ear_clipping.py:27]: This one should stay as-is

File: docs/api/mesh/tessellation.rst [hunk 11]
```diff
@@ -0,0 +1,53 @@
+Tessellation
```
Comment (ktangsali) [inline docs/api/mesh/tessellation.rst:1]: I think it would be good to call out, somewhere that this Tessellation is different from the older Sym's (now deprecated) Tessellation interface. https://github.com/NVIDIA/physicsnemo-sym/blob/main/physicsnemo/sym/geometry/tessellation.py 

Not a big issue for the new users, but for old users, they might not get the context.

File: physicsnemo/mesh/tessellation/_ear_clipping.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def _signed_area2(a: torch.Tensor, b: torch.Tensor, c: torch.Tensor) -> torch.Tensor:
+    """Twice the signed area of 2D triangle ``(a, b, c)`` (positive = CCW)."""
+    return (b[..., 0] - a[..., 0]) * (c[..., 1] - a[..., 1]) - (
+        b[..., 1] - a[..., 1]
+    ) * (c[..., 0] - a[..., 0])
+
+
+def _point_in_triangle(
```
Comment (ktangsali) [inline physicsnemo/mesh/tessellation/_ear_clipping.py:212]: I think the ear test should treat non-corner vertices on the candidate ear boundary as blockers. Right now `_point_in_triangle()` uses strict `> 0` checks, so a vertex exactly on an ear edge is classified as outside. For a simple concave polygon like `[(0,0), (3,0), (3,2), (1.5,1), (0,2)]`, the candidate ear `(4, 0, 1)` has vertex `3` on diagonal `(4, 1)`, but it can still be clipped, producing overlapping triangles and an unsigned area larger than the polygon area.

Could we make the containment test boundary-inclusive, ideally with a small tolerance, and add this as a regression test?

File: physicsnemo/mesh/tessellation/triangulate.py [hunk 13]
```diff
@@ -0,0 +1,465 @@
...
+    counts = polygons.counts
+    # Cheap structural validation, off the torch.compile path (each check is a
+    # host sync that would otherwise force a graph break).
+    if not torch.compiler.is_compiling() and counts.numel() > 0:
+        if bool((counts < 3).any()):
+            raise ValueError(
+                f"Every polygon needs >= 3 vertices to triangulate; got a "
+                f"polygon with {int(counts.min())} vertices."
+            )
+        if bool((polygons.indices >= points.shape[0]).any()):
```
Comment (ktangsali) [inline physicsnemo/mesh/tessellation/triangulate.py:219]: What happens if we have negative indices? Should we add a `polygons.indices < 0).any()` check?

File: docs/api/mesh/core.rst [hunk 14]
```diff
@@ -19,6 +19,9 @@ available both as ``Mesh`` methods and as standalone functions in the
 corresponding submodules. The methods are thin wrappers that pass ``self`` to
 the standalone functions.
 
+To construct a triangle mesh from a non-simplicial polygon soup, use
```
Comment (ktangsali) [inline docs/api/mesh/core.rst]: Minor terminology question: is there a reason to use "polygon soup" instead of something like "polygonal surface mesh" or "surface mesh with polygonal cells"? I understand "polygon soup" is common in computer graphics and may be technically accurate here because the input does not require full mesh topology, but CAE/CFD users may be less familiar with the term. Maybe define it once, or use the CAE-facing term in docs and keep "polygon soup" where the lack of topology is important.

Comment (peterdsharpe) [reply docs/api/mesh/core.rst]: Defined this inline.

File: physicsnemo/mesh/tessellation/triangulate.py (hunk 13 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/tessellation/triangulate.py:219]: Sure, will add

File: physicsnemo/mesh/tessellation/_ear_clipping.py (hunk 12 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/tessellation/_ear_clipping.py:212]: Sure, will do

File: docs/api/mesh/tessellation.rst (hunk 11 above)
Comment (peterdsharpe) [reply docs/api/mesh/tessellation.rst:1]: Added reference to address

File: physicsnemo/mesh/tessellation/triangulate.py [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    next_pos = starts + (local + 1) % valence
+    return poly_id, prev_pos, next_pos
+
+
+def _to_3d(points: torch.Tensor) -> torch.Tensor:
+    """Embed 2D points in 3D (z = 0) so cross products are well-defined."""
+    if points.shape[-1] == 3:
+        return points
+    pad = points.new_zeros((points.shape[0], 3 - points.shape[-1]))
+    return torch.cat([points, pad], dim=-1)
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/tessellation/triangulate.py]: Already addressed

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1711/commits/cb536486f5862499c53417af145b5bf0d7f5da59

**Review by ktangsali — APPROVED** [review]
LGTM. Thanks for addressing my comments.

File: physicsnemo/mesh/tessellation/triangulate.py [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    normals.index_add_(0, poly_id, edge_cross)
+    return normals
+
+
+def _convex_mask(
+    points: Float[torch.Tensor, "n_points 3"],
+    polygons: Adjacency,
+    normals: Float[torch.Tensor, "n_polygons 3"],
+) -> Int[torch.Tensor, " n_polygons"]:
+    """Boolean mask, ``True`` where a polygon is convex (or degenerate).
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/mesh/tessellation/triangulate.py]: fixed
